File: app/src/test/java/io/intrepid/contest/screens/conteststatus/resultsavailable/ResultsAvailablePresenterTest.java
package io.intrepid.contest.screens.conteststatus.resultsavailable;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.junit.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

import io.intrepid.contest.models.Contest;
import io.intrepid.contest.rest.ContestWrapper;
import io.intrepid.contest.screens.conteststatus.resultsavailable.ResultsAvailableContract.Presenter;
import io.intrepid.contest.screens.conteststatus.resultsavailable.ResultsAvailableContract.View;
import io.intrepid.contest.testutils.BasePresenterTest;
import io.reactivex.functions.Consumer;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class ResultsAvailablePresenterTest extends BasePresenterTest<ResultsAvailablePresenter> {
    @Mock
    View mockView;

    private Presenter presenter;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        presenter = new ResultsAvailablePresenter(mockView, testConfiguration);
    }

    @Test
    public void onViewCreatedShouldRequestContestDetails() {
        presenter.onViewCreated();
        verify(mockView).requestContestDetails(any(), any());
    }

    @Test
    public void onViewCreatedShouldShowContestNameWhenApiCallDoesNotThrowError() {
        ArgumentCaptor<Consumer<ContestWrapper>> captor = ArgumentCaptor.forClass(Consumer.class);

        doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                ContestWrapper response = new ContestWrapper(new Contest());
                response.contest.setTitle("Contest title");
                captor.getValue().accept(response);
                return null;
            }
        }).when(mockView).requestContestDetails(captor.capture(), any());

        presenter.onViewCreated();
        testConfiguration.triggerRxSchedulers();

        verify(mockView).showContestName(any(String.class));
        verify(mockView, never()).showMessage(any(int.class));
    }

    @Test
    public void onViewCreatedShouldShowApiErrorMessageWhenApiCallThrowsError() {
        ArgumentCaptor<Consumer<Throwable>> captor = ArgumentCaptor.forClass(Consumer.class);

        doAnswer(new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                captor.getValue().accept(new Throwable());
                return null;
            }
        }).when(mockView).requestContestDetails(any(), captor.capture());

        presenter.onViewCreated();
        testConfiguration.triggerRxSchedulers();

        verify(mockView).showMessage(any(int.class));
    }

    @Test
    public void onViewResultsButtonClickedShouldShowResultsPage() {
        presenter.onViewResultsButtonClicked();
        verify(mockView).showResultsPage();
    }
}

File: app/src/test/java/io/intrepid/contest/screens/conteststatus/resultsavailable/ResultsAvailablePresenterTest.kt
package io.intrepid.contest.screens.conteststatus.resultsavailable

import io.intrepid.contest.models.Contest
import io.intrepid.contest.rest.ContestWrapper
import io.intrepid.contest.screens.conteststatus.resultsavailable.ResultsAvailableContract.View
import io.intrepid.contest.testutils.BasePresenterTest
import io.reactivex.functions.Consumer
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class ResultsAvailablePresenterTest : BasePresenterTest<ResultsAvailablePresenter>() {
    @Mock
    private lateinit var mockView: View

    /**
     * Extension Function to allow generics in argument captor creation
     */
    inline fun <reified T : Any> argumentCaptor() = ArgumentCaptor.forClass(T::class.java)!!

    @Before
    fun setup() {
        presenter = ResultsAvailablePresenter(mockView, testConfiguration)
    }

    @Test
    fun onViewCreatedShouldRequestContestDetails() {
        presenter.onViewCreated()
        verify<View>(mockView).requestContestDetails(any<Consumer<ContestWrapper>>(), any<Consumer<Throwable>>())
    }

    @Test
    fun onViewCreatedShouldShowContestNameWhenApiCallDoesNotThrowError() {
        val captor = argumentCaptor<Consumer<ContestWrapper>>()
        doAnswer {
            val response = ContestWrapper(Contest())
            response.contest.title = "Contest title"
            captor.value.accept(response)
        }.`when`<View>(mockView).requestContestDetails(captor.capture(), any<Consumer<Throwable>>())

        presenter.onViewCreated()
        testConfiguration.triggerRxSchedulers()

        verify<View>(mockView).showContestName(any(String::class.java))
        verify<View>(mockView, never()).showMessage(any(Int::class.javaPrimitiveType!!))
    }

    @Test
    fun onViewCreatedShouldShowApiErrorMessageWhenApiCallThrowsError() {
        val captor = argumentCaptor<Consumer<Throwable>>()
        doAnswer {
            captor.value.accept(Throwable())
        }.`when`<View>(mockView).requestContestDetails(any<Consumer<ContestWrapper>>(), captor.capture())

        presenter.onViewCreated()
        testConfiguration.triggerRxSchedulers()

        verify<View>(mockView).showMessage(any(Int::class.javaPrimitiveType!!))
    }

    @Test
    fun onViewResultsButtonClickedShouldShowResultsPage() {
        presenter.onViewResultsButtonClicked()
        verify<View>(mockView).showResultsPage()
    }
}
